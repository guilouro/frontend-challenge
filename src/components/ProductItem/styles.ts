import styled, { css } from 'styled-components'
import media from 'styled-media-query'

export const Wrapper = styled.article`
  ${({ theme }) => css`
    background-color: ${theme.colors.mainBg};
    padding: ${theme.spacings.small} ${theme.spacings.xsmall};
    display: grid;
    gap: 1rem;
    grid-template-columns: 14.8rem 1fr;
    grid-template-areas:
      'image info'
      'button button';

    ${media.greaterThan('medium')`
      padding: ${theme.spacings.small};
      grid-template-columns: 20rem 1fr;
      grid-template-areas:
        'image info'
        'image button';
    `}
  `}
`

export const Image = styled.div`
  grid-area: image;

  ${media.lessThan('medium')`
    max-width: 14.8rem;
  `}
`

export const Info = styled.div`
  grid-area: info;
`

export const ButtonWrapper = styled.div`
  grid-area: button;
`

export const Title = styled.h3`
  ${({ theme }) => css`
    font-size: ${theme.font.sizes.small};
    font-weight: ${theme.font.bold};
    color: ${theme.colors.text};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xlarge};
      font-weight: ${theme.font.normal};
    `}
  `}
`

export const PriceLabel = styled.span`
  display: none;

  ${media.greaterThan('medium')`
    display: block;
  `}
`

export const Item = styled.div`
  ${({ theme }) => css`
    display: flex;
    justify-content: space-between;
    font-size: ${theme.font.sizes.xsmall};
    font-weight: ${theme.font.normal};
    color: ${theme.colors.textLight};
    margin: ${theme.spacings.xsmall} 0;

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.small};
    `}
  `}
`
