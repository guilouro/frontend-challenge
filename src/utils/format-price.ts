export default function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('de-DE', {
    style: 'currency',
    currency: 'EUR'
  }).format(price)
}
