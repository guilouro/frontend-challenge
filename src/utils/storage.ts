import { CartObject } from 'hooks/use-cart'

const APP_KEY = 'BARKYN_CART'

export function getStorageItem() {
  if (typeof window === 'undefined') return

  const data = window.localStorage.getItem(APP_KEY)
  return JSON.parse(data || '[]')
}

export function setStorageItem(value: CartObject) {
  if (typeof window === 'undefined') return

  const data = JSON.stringify(value)
  return window.localStorage.setItem(APP_KEY, data)
}
