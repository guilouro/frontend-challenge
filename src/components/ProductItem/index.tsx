import { memo } from 'react'
import Image from 'next/image'
import Button from 'components/Button'
import Review, { ReviewProps } from 'components/Review'
import PromotionalPrice, {
  PromotionalPriceProps
} from 'components/PromotionalPrice'

import * as S from './styles'

export type ProductItemProps = {
  id: string
  title: string
  packageInfo: string
  image: string
  onClickButton: (event: React.MouseEvent<HTMLElement>) => void
} & ReviewProps &
  PromotionalPriceProps

const ProductItem = ({
  title,
  price,
  packageInfo,
  review,
  image,
  onClickButton,
  promotionalPrice = 0
}: ProductItemProps) => (
  <S.Wrapper data-cy="product-item">
    <S.Image>
      <Image src={image} alt={title} width={200} height={200} loading="lazy" />
    </S.Image>
    <S.Info>
      <S.Title>{title}</S.Title>
      <S.Item>
        <S.PriceLabel>Price</S.PriceLabel>
        <PromotionalPrice promotionalPrice={promotionalPrice} price={price} />
      </S.Item>
      <S.Item>
        <span>Package</span>
        <span>{packageInfo}</span>
      </S.Item>
      <S.Item>
        <span>Review</span>
        <span>
          <Review review={review} />
        </span>
      </S.Item>
    </S.Info>
    <S.ButtonWrapper>
      <Button onClick={onClickButton}>Add to cart</Button>
    </S.ButtonWrapper>
  </S.Wrapper>
)

export default memo(ProductItem)
