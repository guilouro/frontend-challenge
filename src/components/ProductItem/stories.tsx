import { Story, Meta } from '@storybook/react'
import ProductItem, { ProductItemProps } from '.'

export default {
  title: 'ProductItem',
  component: ProductItem,
  args: {
    title: 'Barkyn Cannabis Oil para c√£es',
    image: '/img/products/oil.png',
    price: 2.29,
    promotionalPrice: 2.24,
    packageInfo: '170g',
    review: 3
  },
  argTypes: {
    review: {
      control: { type: 'range', min: 0, max: 5 }
    }
  }
} as Meta

export const Default: Story<ProductItemProps> = (args) => (
  <div style={{ maxWidth: 570 }}>
    <ProductItem {...args} />
  </div>
)
