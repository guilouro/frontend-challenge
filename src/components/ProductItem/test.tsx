import 'intl.mock'
import { render, screen } from 'utils/test-utils'

import ProductItem from '.'

const props = {
  title: 'Barkyn Cannabis Oil para c√£es',
  image: '/img/products/oil.png',
  price: 2.29,
  packageInfo: '170g',
  review: 3
}

describe('<ProductItem />', () => {
  it('should render correctly', () => {
    const { container } = render(<ProductItem {...props} />)

    expect(
      screen.getByRole('heading', { name: props.title })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('button', { name: /add to cart/i })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
