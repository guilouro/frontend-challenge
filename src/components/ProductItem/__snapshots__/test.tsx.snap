// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> should render correctly 1`] = `
.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  background-color: #5f2eff;
  border-radius: 1.6rem;
  border: 0;
  color: #fff;
  cursor: pointer;
  font-family: Azo sans,sans-serif;
  font-size: 1.4rem;
  min-height: 4rem;
  padding: 0.8rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.2s ease-in-out background-color;
  transition: 0.2s ease-in-out background-color;
}

.c10:hover {
  background-color: #5400db;
}

.c10:disabled {
  background-color: #6c6b76;
}

.c7 {
  color: #f1c40f;
}

.c8 {
  color: #e7e7e9;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  font-size: 1.4rem;
  font-weight: 600;
  color: #5f2eff;
}

.c0 {
  background-color: #fafafa;
  padding: 2.4rem 1.6rem;
  display: grid;
  gap: 1rem;
  grid-template-columns: 14.8rem 1fr;
  grid-template-areas: 'image info' 'button button';
}

.c1 {
  grid-area: image;
}

.c2 {
  grid-area: info;
}

.c9 {
  grid-area: button;
}

.c3 {
  font-size: 1.4rem;
  font-weight: 600;
  color: #383743;
}

.c5 {
  display: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 1.2rem;
  font-weight: 400;
  color: #6c6b76;
  margin: 1.6rem 0;
}

@media (min-width:768px) {
  .c0 {
    padding: 2.4rem;
    grid-template-columns: 20rem 1fr;
    grid-template-areas: 'image info' 'image button';
  }
}

@media (max-width:768px) {
  .c1 {
    max-width: 14.8rem;
  }
}

@media (min-width:768px) {
  .c3 {
    font-size: 2.0rem;
    font-weight: 400;
  }
}

@media (min-width:768px) {
  .c5 {
    display: block;
  }
}

@media (min-width:768px) {
  .c4 {
    font-size: 1.4rem;
  }
}

<article
  class="c0"
>
  <div
    class="c1"
  >
    <div
      style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
    >
      <div
        style="box-sizing: border-box; display: block; max-width: 100%;"
      >
        <img
          alt=""
          aria-hidden="true"
          src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjAwIiBoZWlnaHQ9IjIwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
          style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
        />
      </div>
      <img
        alt="Barkyn Cannabis Oil para cães"
        data-nimg="intrinsic"
        decoding="async"
        src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
        style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
      />
      <noscript />
    </div>
  </div>
  <div
    class="c2"
  >
    <h3
      class="c3"
    >
      Barkyn Cannabis Oil para cães
    </h3>
    <div
      class="c4"
    >
      <span
        class="c5"
      >
        Price
      </span>
      <div>
        <div
          class="c6"
        >
          2.29
        </div>
      </div>
    </div>
    <div
      class="c4"
    >
      <span>
        Package
      </span>
      <span>
        170g
      </span>
    </div>
    <div
      class="c4"
    >
      <span>
        Review
      </span>
      <span>
        <div
          aria-label="3 stars of 5"
          role="img"
        >
          <svg
            class="c7"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="m0 0h24v24h-24z"
              />
              <path
                d="m12 18-5.26 2.765a1 1 0 0 1 -1.45-1.054l1.004-5.857-4.256-4.148a1 1 0 0 1 .554-1.706l5.881-.854 2.63-5.329a1 1 0 0 1 1.794 0l2.63 5.329 5.88.854a1 1 0 0 1 .555 1.706l-4.256 4.148 1.005 5.857a1 1 0 0 1 -1.451 1.054z"
                fill="currentColor"
              />
            </g>
          </svg>
          <svg
            class="c7"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="m0 0h24v24h-24z"
              />
              <path
                d="m12 18-5.26 2.765a1 1 0 0 1 -1.45-1.054l1.004-5.857-4.256-4.148a1 1 0 0 1 .554-1.706l5.881-.854 2.63-5.329a1 1 0 0 1 1.794 0l2.63 5.329 5.88.854a1 1 0 0 1 .555 1.706l-4.256 4.148 1.005 5.857a1 1 0 0 1 -1.451 1.054z"
                fill="currentColor"
              />
            </g>
          </svg>
          <svg
            class="c7"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="m0 0h24v24h-24z"
              />
              <path
                d="m12 18-5.26 2.765a1 1 0 0 1 -1.45-1.054l1.004-5.857-4.256-4.148a1 1 0 0 1 .554-1.706l5.881-.854 2.63-5.329a1 1 0 0 1 1.794 0l2.63 5.329 5.88.854a1 1 0 0 1 .555 1.706l-4.256 4.148 1.005 5.857a1 1 0 0 1 -1.451 1.054z"
                fill="currentColor"
              />
            </g>
          </svg>
          <svg
            class="c8"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="m0 0h24v24h-24z"
              />
              <path
                d="m12 18-5.26 2.765a1 1 0 0 1 -1.45-1.054l1.004-5.857-4.256-4.148a1 1 0 0 1 .554-1.706l5.881-.854 2.63-5.329a1 1 0 0 1 1.794 0l2.63 5.329 5.88.854a1 1 0 0 1 .555 1.706l-4.256 4.148 1.005 5.857a1 1 0 0 1 -1.451 1.054z"
                fill="currentColor"
              />
            </g>
          </svg>
          <svg
            class="c8"
            height="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="none"
              fill-rule="evenodd"
            >
              <path
                d="m0 0h24v24h-24z"
              />
              <path
                d="m12 18-5.26 2.765a1 1 0 0 1 -1.45-1.054l1.004-5.857-4.256-4.148a1 1 0 0 1 .554-1.706l5.881-.854 2.63-5.329a1 1 0 0 1 1.794 0l2.63 5.329 5.88.854a1 1 0 0 1 .555 1.706l-4.256 4.148 1.005 5.857a1 1 0 0 1 -1.451 1.054z"
                fill="currentColor"
              />
            </g>
          </svg>
        </div>
      </span>
    </div>
  </div>
  <div
    class="c9"
  >
    <button
      class="c10"
    >
      <span>
        Add to cart
      </span>
    </button>
  </div>
</article>
`
